### PR TITLE
Add lines to plot to main

### PR DESCRIPTION
**Pull request by aronnoordam**

### CODE REVIEW

File: stem/utils.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
     from stem.mesh import Element, Mesh
     from stem.geometry import Geometry
 
+NUMBER_TYPES = (int, float, np.int64, np.float64)
+
```
Comment (brunozc) [inline stem/utils.py:15]: """
TypeAlias:
    - NUMBER_TYPES: Union[int, float, np.int64, np.float64]
"""

File: stem/utils.py [hunk 2]
```diff
@@ -376,6 +378,47 @@ def create_sigmoid_tiny_expr(start_time: float, dt_slope: float, initial_value:
             return (f"(1 / (1 + e^(-{beta} * (t - {dt_slope} / 2 - {start_time})))) * ({final_value} - {initial_value})"
                     f"+ {initial_value}")
 
+    @staticmethod
+    def create_box_tiny_expr(transition_parameter: float,
```
Comment (brunozc) [inline stem/utils.py:382]: already reviewed in first PR

File: stem/utils.py (hunk 1 above)
Comment (brunozc) [reply stem/utils.py:15]: Arent you missing: numpy.float96 numpy.float128

**Review by brunozc — CHANGES_REQUESTED** [review]
I havent reviewed the html nor the json files

File: stem/geometry.py [hunk 3]
```diff
@@ -502,3 +505,183 @@ def create_geometry_from_gmsh_group(cls, geo_data: Dict[str, Any], group_name: s
...
+        """
+        Calculate the centre of mass of a surface.
+
+        Args:
+            - surface_id (int): The id of the surface.
+
+        Returns:
+            - npty.NDArray[np.float64]: The coordinates of the centre of mass of the surface.
+        """
+        # initialize centre of mass and circumference
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/geometry.py:596]: maybe the naming is wrong, but with the centroid, if you have the following figure:

image

the centroid will be somewhere around the black circle. While i want to get the location of the blue circle.
For this i gave "mass" to the lines inbetween points. With this the centre of mass is calculated

File: stem/geometry.py [hunk 4]
```diff
@@ -502,3 +505,183 @@ def create_geometry_from_gmsh_group(cls, geo_data: Dict[str, Any], group_name: s
...
+        # calculate the normal vector of the surface
+        normal_vector = np.cross(non_collinear_points[1] - non_collinear_points[0],
+                                 non_collinear_points[2] - non_collinear_points[0])
+
+        # calculate the dot product of the cross products and the normal vector in order to determine the sign of the
+        # area increment
+        cross_dot_normal = np.dot(cross_products, normal_vector)
+
+        # calculate the signed area of the surface
+        area: float = norms.dot(np.sign(cross_dot_normal)) * 0.5
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/geometry.py:654]: i think this was for mypy, but i will check

File: stem/geometry.py [hunk 5]
```diff
@@ -502,3 +505,183 @@ def create_geometry_from_gmsh_group(cls, geo_data: Dict[str, Any], group_name: s
...
+
+        # return the absolute value of the area
+        return abs(area)
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/geometry.py:657]: i think this was for mypy, but i will check

File: stem/geometry.py [hunk 6]
```diff
@@ -502,3 +505,183 @@ def create_geometry_from_gmsh_group(cls, geo_data: Dict[str, Any], group_name: s
...
+
+    def calculate_centre_of_mass_volume(self, volume_id: int) -> npty.NDArray[np.float64]:
+        """
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/geometry.py:660]: see comment at surface

File: stem/mesh.py [hunk 7]
```diff
@@ -285,3 +286,37 @@ def create_mesh_from_gmsh_group(cls, mesh_data: Dict[str, Any], group_name: str)
         mesh.elements = elements
 
         return mesh
+
+    def calculate_centroids(self) -> npty.NDArray[np.float64]:
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/mesh.py:290]: but in this function it also gets the coordinates from the elements.

This part is different in mesh.py and geometry.py

File: stem/model.py [hunk 8]
```diff
@@ -503,105 +472,13 @@ def add_load_on_line_model_part(self, model_part_name: str, load_parameters: Loa
...
-
         # add load parameters to model part
         model_part.parameters = load_parameters
+        # validate the input
+        model_part.validate_input()
 
         self.process_model_parts.append(model_part)
 
-    @staticmethod
-    def validate_coordinates(coordinates: Union[Sequence[Sequence[float]], npty.NDArray[np.float64]]):
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:536]: moved to utils

File: stem/model.py [hunk 9]
```diff
@@ -503,105 +472,13 @@ def add_load_on_line_model_part(self, model_part_name: str, load_parameters: Loa
...
-
-        # check if coordinates are real numbers
-        for coordinate in coordinates:
-            for i in coordinate:
-                if not isinstance(i, NUMBER_TYPES) or np.isnan(i) or np.isinf(i):
-                    raise ValueError(f"Coordinates should be a sequence of sequence of real numbers, "
-                                     f"but {i} was given.")
-
-    @staticmethod
-    def __validate_moving_load_parameters(coordinates: Sequence[Sequence[float]],
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:568]: moved to utils

File: stem/model.py [hunk 10]
```diff
@@ -721,8 +600,8 @@ def add_output_settings_by_coordinates(self,
         # TODO: add validation for sequential pair of points to lie on the an existing geometry surface.
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:600]: i dont think its done. Validate coordinates just checks if its 2d,3d, and if its nan or infinite

[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:600]: https://github.com/StemVibrations/STEM/issues/248

File: stem/model.py [hunk 11]
```diff
@@ -938,11 +802,14 @@ def generate_mesh(self,
     def __post_mesh(self):
         """
         Function to be called after the mesh is generated and finalised.
-            - initialise field parameters (e.g., random fields).
-            - adjust the elements for the spring damper parts.
+            - initialise field parameters (e.g., random fields). |
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:805]: I think its |

[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:805]: https://rest-sphinx-memo.readthedocs.io/en/latest/ReST.html#line-blocks

File: stem/model.py [hunk 12]
```diff
@@ -1060,23 +930,22 @@ def __get_line_string_end_nodes(self, model_part: ModelPart) -> List[List[int]]:
...
 
         # find end nodes
-        end_nodes = self.__find_end_nodes_of_line_strings(model_part.mesh)
+        end_nodes = set(self.__find_end_nodes_of_line_strings(model_part.mesh))
         # find the node ids corresponding to the geometry points
-        node_ids_at_geometry_points = Utils.find_node_ids_close_to_geometry_nodes(mesh=model_part.mesh,
-                                                                                  geometry=model_part.geometry,
-                                                                                  eps=1e-06)
+        node_ids_at_geometry_points = set(
+            Utils.find_node_ids_close_to_geometry_nodes(mesh=model_part.mesh, geometry=model_part.geometry, eps=1e-06))
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:936]: makes sense, there already is globals.py

File: stem/model.py [hunk 13]
```diff
@@ -1176,9 +1021,8 @@ def __find_next_node_along_line_elements(start_node_id: int, remaining_element_i
         """
 
         # check if all elements are line elements
-        for element in line_elements.values():
-            if element.element_type != "LINE_2N" and element.element_type != "LINE_3N":
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:1180]: https://github.com/StemVibrations/STEM/issues/249

File: stem/model.py [hunk 14]
```diff
@@ -1213,24 +1057,6 @@ def __find_next_node_along_line_elements(start_node_id: int, remaining_element_i
...
-
-        """
-        if model_part.mesh is not None:
-            return np.array([el.node_ids for el in model_part.mesh.elements.values()])
-        else:
-            return np.array([])
-
     def __find_matching_body_elements_for_process_model_part(self, process_model_part: ModelPart) \
             -> List[Tuple[Element, Element]]:
         """
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:1062]: finds

File: stem/model.py [hunk 15]
```diff
@@ -1682,3 +1436,90 @@ def set_element_size_of_group(self, element_size: float, group_name: str):
...
+        if isinstance(from_model_part, BodyModelPart) and isinstance(new_parameters, get_args(Material)):
+
+            # check if the new parameters are of the same type as the existing material
+            if not isinstance(new_parameters, from_model_part.material.__class__):
+                raise ValueError("New parameters must have the same material type as in the original "
+                                 "body model part.")
+
+            # create a new body model part
+            new_model_part = BodyModelPart(name=to_model_part_name)
+            new_model_part.material = new_parameters  # type: ignore
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:1478]: its for mypy or yapf or flake8

File: stem/utils.py (hunk 1 above)
Comment (aronnoordam) [reply stem/utils.py:15]: are those relevant?

File: stem/model.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/model.py:936]: https://github.com/StemVibrations/STEM/issues/249

**Review by brunozc — APPROVED** [review]

**Pull request merged**